storage_service/common/datatable_utils.py

```diff
@@ -0,0 +1,159 @@
+            )
+            if sorting_column_index is not None:
+                is_sortable = (
+                    query_dict.get("bSortable_{}".format(sorting_column_index))
+                    == "true"
+                )
+                if is_sortable:
+                    sort_direction = query_dict.get("sSortDir_0", "asc")
+                    result = {
+                        "index": sorting_column_index,
+                        "direction": sort_direction,
+                    }
+        return result
+
+    def parse_datatable_parameters(self, query_dict):
+        return {
+            # query for text based search
+            "search": query_dict.get("sSearch", ""),
+            # options for limiting result display
+            "display_start": self._get_int_parameter(query_dict, "iDisplayStart"),
+            "display_length": self._get_int_parameter(query_dict, "iDisplayLength"),
+            # sorting information
+            "sorting_column": self.sorting_column(query_dict),
+            # parameter sent by the client that the server needs
+            # to return back in order to identify each single rendering
+            "echo": self._get_int_parameter(query_dict, "sEcho", -1),
+        }
+
+    def sort(self, queryset):
+        sorting_column = self.params["sorting_column"]
+        if not sorting_column or sorting_column.get("index") is None:
+            return queryset
+        # the columns represented by these indexes can be sorted
+        # through the queryset using the order_by method
+        ORDER_BY_MAPPING = {
+            0: "uuid",
+            1: "origin_pipeline__description",
+            3: "size",
+            5: "replicated_package__uuid",
+        }
+        # these columns instead can't be sorted directly with the
+        # queryset so the queryset needs to be converted into a list
+        # and then sorted using helper methods
+        SORT_KEY_HELPERS_MAPPING = {
+            2: self.sort_by_full_path_key,
+            4: self.sort_by_package_type_key,
+            6: self.sort_by_status_key,
+            7: self.sort_by_fixity_date_key,
+            8: self.sort_by_fixity_status_key,
+        }
+        sort_descending = sorting_column.get("direction") == "desc"
+        if sorting_column["index"] in ORDER_BY_MAPPING:
+            field = ORDER_BY_MAPPING[sorting_column["index"]]
+            if sort_descending:
+                field = "-{}".format(field)
+            return queryset.order_by(field)
+        elif sorting_column["index"] in SORT_KEY_HELPERS_MAPPING:
+            key = SORT_KEY_HELPERS_MAPPING[sorting_column["index"]]
+            queryset = sorted(list(queryset), key=key, reverse=sort_descending)
+        else:
+            return queryset
+
+    def get_packages(self, queryset):
+        result = self.sort(queryset)
+        display_start = self.params["display_start"]
+        display_length = self.params["display_length"]
+        if display_start and display_length:
```

Comment:
`display_start` is 0 on the first page, so this will return all results for that page 😁. I think we need to check for `is not None`.

Reply:
Looks like I need to improve the tests too!

Reply:
In https://github.com/artefactual/archivematica-storage-service/pull/470/commits/d291a7b18cd7222b327d34c7098cdaa23852c822 I set a default number of rows to display and changed the condition to verify if some result exists.